### PR TITLE
use site-url setting when opening links via JS

### LINKED ISSUE #11684: Opening in new tab missing initial path component behind proxy
**Describe the bug**
When metabase is hosted behind a proxy with a path, Ctrl-Clicking a question from a dashboard opens a new tab with /question/:id instead of /metabase/question/:id

Related: #7841, #5168, #6061

Seems that this functionality was introduced in #8860 - I'm not familiar enough with the codebase to say what exactly, but *something* in there needs to prepend URL with the system's root URL before opening a new tab.

**To Reproduce**
1. Host metabase at a path (we're hosting it at labs.waytohealth.org/metabase)
2. Add a question to a dashboard
3. View the dashboard
4. Ctrl-click the question title

**Expected behavior**
A new tab opens at https://labs.waytohealth.org/metabase/question/288

**Actual behavior**
A new tab opens at https://labs.waytohealth.org/question/288

**Information about your Metabase Installation:**
```
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.117 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "1.8.0_232-b09",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_232",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "25.232-b09",
    "os.name": "Linux",
    "os.version": "3.10.0-1062.9.1.el7.x86_64",
    "user.language": "en",
    "user.timezone": "America/New_York"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "mysql",
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "mysql",
    "run-mode": "prod",
    "version": {
      "tag": "v0.34.0",
      "date": "2019-12-19",
      "branch": "release-0.34.x",
      "hash": "1335215"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**
Mildly annoying

### LINKED ISSUE #12580: Create your metric Preview button goes to the wrong place behind a reverse proxy
Create your metric Preview button goes to the wrong place behind a reverse proxy, where Metabase is non-root / sub-directory:

image

It redirects to `{HOSTNAME}/question#klsjdflksdjf...`

Instead of `{HOSTNAME}/metabase/question#klsjdflksdjf...`

_Originally from https://github.com/metabase/metabase/issues/6061#issuecomment-393910209_

Related #10743 #11684

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### LINKED ISSUE #13602: Click Behavior - Link to Dashboard / Question does not respect SITE_URL setting in Admin Panel
**Describe the bug**
Click behavior -> Link to a dashboard / question strips the path of the URL. It does not respect the `SITE_URL` settings in Admin page.

For example, with setting `SITE_URL`: http://example.com/metabase, the click behavior brings the user to http://example.com/dashboard/18. However, it should bring the user to http://example.com/metabase/dashboard/18

**To Reproduce**
1. Set up click behavior to a column, Link to any other dashboard or question.
2. Save the dashboard.
3. Click on the column.
3. The dashboard redirects to the URL with the path stripped.

**Expected behavior**
It should append the `/dashboard` to `SITE_URL`.

**Information about your Metabase Installation:**

You can get this information by going to Admin -> Troubleshooting.

- Metabase version: v0.37.0.2
- Metabase hosting environment: Jar-file on RHEL 7.5, reverse proxied by NGINX at the path /metabase
- Metabase internal database: PostgreSQL 

**Severity**
It causes confusion and complains from users when the functionality was provided but does not work in our case.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### PR DESCRIPTION
**Pull request by daltojohnso**
Fixes #13602 

We weren't using the `site-url` setting in places where we relied on the `open` function from `metabase/lib/dom`. I've changed it so that intra-app navigation uses the router via `push`. This is nice because it avoids a site reload. In scenarios where we open a new tab/window, I'm appending the `site-url` when `openInBlankWindow` is given a path.

**Testing**

Setup
- Run a proxy so that you can open up an instance of metabase at a url like `localhost:3100/metabase`. I'm using nginx to do this. You can use the config for nginx found in this PR: https://github.com/metabase/metabase/pull/4740
- Go to `/admin/settings/general` and update the site url input to `localhost:3100/metabase`

Click behavior

Question link:

https://user-images.githubusercontent.com/13057258/154768273-514e1453-661e-4ddd-8f37-6f2488d2cb6d.mov

Dashboard link:

https://user-images.githubusercontent.com/13057258/154768583-d16e3116-62a5-441f-9574-afff81e1e593.mov

Clicking a dashcard title while holding down ctrl or cmd opens the correct link:

https://user-images.githubusercontent.com/13057258/154767725-df4c268c-342a-478f-bf16-2c27de81b2bb.mov

QB viz click -- cmd + clicking on a drilldown option opens the correct link:

https://user-images.githubusercontent.com/13057258/154767664-e472c876-49e8-4f5e-80d7-4ef8f4c34660.mov

### CODE REVIEW

File: frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx [hunk 1]
```diff
@@ -112,12 +112,12 @@ export default class Dashboard extends Component {
     });
   }
 
-  UNSAFE_componentWillReceiveProps(nextProps) {
-    if (this.props.dashboardId !== nextProps.dashboardId) {
-      this.loadDashboard(nextProps.dashboardId);
+  componentDidUpdate(prevProps) {
```
Comment (daltojohnso) [inline frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx:115]: `UNSAFE_componentWillReceiveProps` was causing a race condition, actually, where `this.props.dashboardId` had changed but `this.props.location` had not.

File: frontend/src/metabase/parameters/components/widgets/ParameterFieldWidget/normalizeValue.js [hunk 2]
```diff
@@ -3,5 +3,5 @@ export function normalizeValue(value) {
     return value;
   }
 
-  return value ? [value] : [];
```
Comment (daltojohnso) [inline frontend/src/metabase/parameters/components/widgets/ParameterFieldWidget/normalizeValue.js:6]: This is unrelated to this issue, but happened to notice it while testing.

**Review by flamber — APPROVED** [review]
This works great! Also fixes #11684.

One problem though:
It's impossible to go to URL destination for anything outside of Metabase on the same domain. Example:
Metabase is on `https://example.com/metabase/` and I have a click behavior URL as `/myapp/` (or `https://example.com/myapp/`), then when clicked it shows "We're a little lost...", which is understandable, since it's trying to send my to `/metabase/myapp/`

I know it's out of scope, but how difficult would it be to fix the couple of related issues, so we have full non-root support? https://github.com/metabase/metabase/issues/13602#issuecomment-717154549

It seems like we're not adding the non-root path, when doing SSO logins - I only tested JWT, but I guess it applies in general. I cannot understand what Clojure is doing https://github.com/metabase/metabase/blob/master/enterprise/backend/src/metabase_enterprise/sso/integrations/jwt.clj, but it doesn't seem to include the non-root and it also defaults to just `/` if no redirect is included.

**Comment by daltojohnso** [comment]
@flamber I think you can set the click behavior url to `../myapp/` and it'll navigate to `https://example.com/myapp/` correctly

I'll work on those other issues in separate PRs because the work is pretty different from this one.

**Comment by daltojohnso** [comment]
Fixes #12580 now too

https://user-images.githubusercontent.com/13057258/155438367-43679fc4-c8b7-415e-9974-312ec287d2a5.mov

**Review by ariya — APPROVED** [review]
Let's ship this :rocket:

**Pull request merged**